File: src/qrs-core-js/src/services/trustService.ts
/**
 * TrustService: pinning, CA roles, attestations, and trust resolution.
 *
 * Trust model (v1): two levels.
 *   - Pinned TCert: the verifier explicitly trusts the TCert.
 *   - CA-issued TCert: a TCert granted CA authority has signed an `attest` statement
 *     about the target TCert.
 *
 * A TCert can be trusted through both mechanisms at once (dual endorsement). The
 * CA is not a special object type — it is an ordinary TCert the verifier has
 * configured with CA authority.
 */
import type { PublicJwk } from '../crypto/jwk.js';
import { QrsAuthorizationError, QrsCryptoError, QrsNotFoundError, QrsValidationError } from '../errors.js';
import { splitTcertId, tcertIdOf } from '../signedObject/signedObject.js';
import { parseSignedObject, verifyParsedSignedObject, type ParsedSignedObject } from '../signedObject/signedObject.js';
import { tcertHashOf } from '../signedObject/signedObject.js';
import type { AlgorithmId, KeyId, TcertId } from '../types.js';
import type { AttestationRecord } from '../storage/stores.js';
import type { ServiceDeps } from './deps.js';
import { buildStatement, parseStatement, verifyStatement } from './statement.js';

export type ComponentState = 'valid' | 'invalid' | 'cannotVerify';

export interface TrustResolution {
  state: ComponentState;
  pinned: boolean;
  ca?: { caTcertId: TcertId; caName?: string };
  message?: string;
}

export interface AttestParams {
  caTcertId: TcertId;
  targetTcertId: TcertId;
  claims?: Record<string, unknown>;
  issuedAt?: number;
}

export interface AddTcertParams extends AttestParams {
  tcertBytes?: Uint8Array;
}

export interface TcertKeyInfo {
  keyId: KeyId;
  algorithm: AlgorithmId;
  publicJwk: PublicJwk;
}

export class TrustService {
  constructor(private readonly deps: ServiceDeps) {}

  /* ---------------- trust configuration ---------------- */

  async pin(tcertId: TcertId): Promise<void> {
    await this.ensureTcert(tcertId);
    await this.deps.trustStore.addPinned(tcertId);
  }

  async unpin(tcertId: TcertId): Promise<void> {
    await this.deps.trustStore.removePinned(tcertId);
  }

  async addCa(tcertId: TcertId): Promise<void> {
    await this.ensureTcert(tcertId);
    await this.deps.trustStore.addCa(tcertId);
  }

  async removeCa(tcertId: TcertId): Promise<void> {
    await this.deps.trustStore.removeCa(tcertId);
  }

  async distrust(tcertId: TcertId): Promise<void> {
    await this.deps.trustStore.addDistrust(tcertId);
  }

  async trustAgain(tcertId: TcertId): Promise<void> {
    await this.deps.trustStore.removeDistrust(tcertId);
  }

  /* ---------------- attestation ---------------- */

  /** A CA signs an attestation about a target TCert (does not modify the target). */
  async attest(params: AttestParams): Promise<{ statementId: string; bytes: Uint8Array }> {
    if (!(await this.deps.trustStore.isCa(params.caTcertId))) {
      throw new QrsAuthorizationError(`TCert ${params.caTcertId} is not configured as a CA`);
    }
    // Prevent duplicate attestations: the same CA must not attest the same target twice.
    const existing = await this.deps.trustStore.getAttestations(params.targetTcertId);
    if (existing.some((r) => r.caTcertId === params.caTcertId)) {
      throw new QrsValidationError(`TCert ${params.targetTcertId} is already attested by ${params.caTcertId}`);
    }
    const ca = await this.keyInfoOf(params.caTcertId);
    const priv = await this.deps.privateKeyStore.load(ca.keyId);
    if (!priv) throw new QrsNotFoundError(`CA private key not available: ${ca.keyId}`);

    const target = splitTcertId(params.targetTcertId);
    // The attestation binds the content hash of the specific TCert being attested.
    const targetBytes = await this.deps.certificateStore.get(params.targetTcertId);
    if (!targetBytes) throw new QrsNotFoundError(`TCert not found: ${params.targetTcertId}`);
    const targetParsed = parseSignedObject(targetBytes);
    const tcertHash = tcertHashOf(targetParsed);
    const issuedAt = params.issuedAt ?? this.deps.clock.now();
    const keyPair = { algorithm: ca.algorithm, publicJwk: ca.publicJwk, privateJwk: priv.privateJwk };
    const built = await buildStatement(
      'attest',
      { kind: 'tcert', keyId: target.keyId, certificateNumber: target.certificateNumber, tcertHash },
      issuedAt,
      { claims: params.claims },
      keyPair,
      this.deps.cryptoRegistry.get(ca.algorithm)
    );
    if (!(await verifyStatement(built.parsed, ca.publicJwk, this.deps.cryptoRegistry.get(ca.algorithm)))) {
      throw new QrsCryptoError('Attestation statement failed verification');
    }
    const record: AttestationRecord = {
      targetTcertId: params.targetTcertId,
      caKeyId: ca.keyId,
      caTcertId: params.caTcertId,
      tcertHash,
      claims: params.claims,
      issuedAt,
      statementBytes: built.bytes,
    };
    await this.deps.trustStore.addAttestation(record);
    return { statementId: built.statementId, bytes: built.bytes };
  }

  /** A CA introduces a TCert into its namespace (optionally storing its bytes first). */
  async addTcert(params: AddTcertParams): Promise<{ statementId: string; bytes: Uint8Array }> {
    if (params.tcertBytes) {
      const parsed = parseSignedObject(params.tcertBytes);
      if (parsed.type !== 'tcert') throw new QrsValidationError('Provided object is not a TCert');
      const tcertId = tcertIdOf(parsed.signerKeyId, parsed.data.certificateNumber as number);
      if (tcertId !== params.targetTcertId) {
        throw new QrsValidationError('Provided TCert bytes do not match the target tcert id');
      }
      await this.deps.certificateStore.save(tcertId, params.tcertBytes);
    }
    return this.attest(params);
  }

  /* ---------------- trust resolution ---------------- */

  /**
   * Resolve trust for a TCert. Returns the strongest trust state available.
   * Pinned TCerts stay valid even if the attesting CA is later revoked (the CA name
   * is simply not shown); a purely CA-issued TCert depends on its CA.
   */
  async resolveTrust(tcertId: TcertId, _parsed?: ParsedSignedObject): Promise<TrustResolution> {
    if (await this.deps.trustStore.isDistrusted(tcertId)) {
      return { state: 'invalid', pinned: false, message: 'TCert is locally distrusted' };
    }
    const pinned = await this.deps.trustStore.isPinned(tcertId);
    const attestations = await this.deps.trustStore.getAttestations(tcertId);
    let ca: { caTcertId: TcertId; caName?: string } | null = null;
    for (const att of attestations) {
      if (await this.deps.revocationStore.getRevokedAttestation(att.targetTcertId, att.caTcertId)) continue;
      if (await this.isValidAttestation(att)) {
        ca = { caTcertId: att.caTcertId, caName: (att.claims?.name as string) ?? undefined };
        break;
      }
    }
    if (pinned) return { state: 'valid', pinned: true, ca: ca ?? undefined };
    if (ca) return { state: 'valid', pinned: false, ca };
    return {
      state: 'cannotVerify',
      pinned: false,
      message: 'no trust path: TCert is neither pinned nor attested by a trusted CA',
    };
  }

  /** True when `caTcertId` is a trusted CA in good standing. */
  async isTrustedCa(caTcertId: TcertId): Promise<boolean> {
    if (await this.deps.trustStore.isDistrusted(caTcertId)) return false;
    if (await this.deps.revocationStore.getRevokedTcert(caTcertId)) return false;
    const pinned = await this.deps.trustStore.isPinned(caTcertId);
    const isCa = await this.deps.trustStore.isCa(caTcertId);
    return pinned || isCa;
  }

  /** Validate an attestation record end-to-end (CA trusted, statement signature valid). */
  async isValidAttestation(att: AttestationRecord): Promise<boolean> {
    const caBytes = await this.deps.certificateStore.get(att.caTcertId);
    if (!caBytes) return false;
    let caParsed: ParsedSignedObject;
    try {
      caParsed = parseSignedObject(caBytes);
    } catch {
      return false;
    }
    if (caParsed.type !== 'tcert') return false;
    const provider = this.deps.cryptoRegistry.get(caParsed.algorithm);
    const caPub = caParsed.data.publicKey as unknown as PublicJwk;
    if (!(await verifyParsedSignedObject(caParsed, provider, caPub))) return false;
    if (!(await this.isTrustedCa(att.caTcertId))) return false;
    if (await this.deps.revocationStore.getRevokedKey(att.caKeyId)) return false;

    let stmt;
    try {
      stmt = parseStatement(att.statementBytes);
    } catch {
      return false;
    }
    if (stmt.action !== 'attest') return false;
    if (stmt.target.kind !== 'tcert') return false;
    if (tcertIdOf(stmt.target.keyId, stmt.target.certificateNumber) !== att.targetTcertId) return false;
    if (stmt.signerKeyId !== att.caKeyId) return false;

    // The attestation must bind the exact content hash of the TCert it is about.
    const targetBytes = await this.deps.certificateStore.get(att.targetTcertId);
    if (!targetBytes) return false;
    let targetParsed: ParsedSignedObject;
    try {
      targetParsed = parseSignedObject(targetBytes);
    } catch {
      return false;
    }
    if (tcertHashOf(targetParsed) !== att.tcertHash) return false;
    if (stmt.target.tcertHash !== att.tcertHash) return false;

    return verifyStatement(stmt.parsed, caPub, provider);
  }

  /* ---------------- helpers ---------------- */

  private async ensureTcert(tcertId: TcertId): Promise<void> {
    const bytes = await this.deps.certificateStore.get(tcertId);
    if (!bytes) throw new QrsNotFoundError(`TCert not found: ${tcertId}`);
  }

  private async keyInfoOf(tcertId: TcertId): Promise<TcertKeyInfo> {
    const bytes = await this.deps.certificateStore.get(tcertId);
    if (!bytes) throw new QrsNotFoundError(`TCert not found: ${tcertId}`);
    const parsed = parseSignedObject(bytes);
    const keyId = parsed.signerKeyId;
    const algorithm = parsed.data.algorithm as AlgorithmId;
    const publicJwk = parsed.data.publicKey as unknown as PublicJwk;
    return { keyId, algorithm, publicJwk };
  }
}

File: qrs-core-python/src/qrs/services/trustService.py
"""TrustService: pinning, CA roles, attestations, and trust resolution.

Trust model (v1): two levels.

- Pinned TCert: the verifier explicitly trusts the TCert.
- CA-issued TCert: a TCert granted CA authority has signed an ``attest``
  statement about the target TCert.

A TCert can be trusted through both mechanisms at once (dual endorsement). The
CA is not a special object type — it is an ordinary TCert the verifier has
configured with CA authority.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from ..crypto.providers import KeyPairMaterial
from ..deps import ServiceDeps
from ..envelope import (
    parse_signed_object,
    split_tcert_id,
    tcert_hash_of,
    tcert_id_of,
    verify_parsed_signed_object,
)
from ..errors import QrsAuthorizationError, QrsCryptoError, QrsNotFoundError, QrsValidationError
from .statement import StatementOptions, StatementTarget, build_statement, parse_statement, verify_statement
from ..storage.interfaces import AttestationRecord

__all__ = ["TrustService", "TrustResolution", "AttestParams", "AddTcertParams"]


@dataclass
class TrustResolution:
    state: str  # 'valid' | 'invalid' | 'cannotVerify'
    pinned: bool = False
    ca: dict[str, str | None] | None = None
    message: str | None = None


@dataclass
class AttestParams:
    ca_tcert_id: str
    target_tcert_id: str
    claims: dict[str, Any] | None = None
    issued_at: int | None = None


@dataclass
class AddTcertParams(AttestParams):
    tcert_bytes: bytes | None = None


class TrustService:
    def __init__(self, deps: ServiceDeps) -> None:
        self._deps = deps

    async def pin(self, tcert_id: str) -> None:
        await self._ensure_tcert(tcert_id)
        await self._deps.trust_store.add_pinned(tcert_id)

    async def unpin(self, tcert_id: str) -> None:
        await self._deps.trust_store.remove_pinned(tcert_id)

    async def add_ca(self, tcert_id: str) -> None:
        await self._ensure_tcert(tcert_id)
        await self._deps.trust_store.add_ca(tcert_id)

    async def remove_ca(self, tcert_id: str) -> None:
        await self._deps.trust_store.remove_ca(tcert_id)

    async def distrust(self, tcert_id: str) -> None:
        await self._deps.trust_store.add_distrusted(tcert_id)

    async def trust_again(self, tcert_id: str) -> None:
        await self._deps.trust_store.remove_distrusted(tcert_id)

    async def attest(self, params: AttestParams) -> dict[str, Any]:
        if not await self._deps.trust_store.is_ca(params.ca_tcert_id):
            raise QrsAuthorizationError(f"TCert {params.ca_tcert_id} is not configured as a CA")
        # Prevent duplicate attestations: the same CA must not attest the same target twice.
        existing = await self._deps.trust_store.get_attestations(params.target_tcert_id)
        if any(record.ca_tcert_id == params.ca_tcert_id for record in existing):
            raise QrsValidationError(
                f"TCert {params.target_tcert_id} is already attested by {params.ca_tcert_id}"
            )
        ca = await self._key_info_of(params.ca_tcert_id)
        priv = await self._deps.private_key_store.load(ca["key_id"])
        if not priv:
            raise QrsNotFoundError(f"CA private key not available: {ca['key_id']}")

        target = split_tcert_id(params.target_tcert_id)
        # The attestation binds the content hash of the specific TCert being attested.
        target_bytes = await self._deps.certificate_store.get(params.target_tcert_id)
        if not target_bytes:
            raise QrsNotFoundError(f"TCert not found: {params.target_tcert_id}")
        target_parsed = parse_signed_object(target_bytes)
        tcert_hash = tcert_hash_of(target_parsed)
        issued_at = params.issued_at if params.issued_at is not None else self._deps.clock.now()
        key_pair = KeyPairMaterial(
            algorithm=ca["algorithm"],
            public_jwk=ca["public_jwk"],
            private_jwk=priv["private_jwk"],
        )
        built = build_statement(
            "attest",
            StatementTarget(
                kind="tcert",
                key_id=target[0],
                certificate_number=target[1],
                tcert_hash=tcert_hash,
            ),
            issued_at,
            StatementOptions(claims=params.claims),
            key_pair,
            self._deps.crypto_registry.get(ca["algorithm"]),
        )
        if not verify_statement(built.parsed, ca["public_jwk"], self._deps.crypto_registry.get(ca["algorithm"])):
            raise QrsCryptoError("Attestation statement failed verification")
        record = AttestationRecord(
            target_tcert_id=params.target_tcert_id,
            ca_tcert_id=params.ca_tcert_id,
            ca_key_id=ca["key_id"],
            tcert_hash=tcert_hash,
            claims=params.claims,
            issued_at=issued_at,
            statement_bytes=built.bytes,
        )
        await self._deps.trust_store.add_attestation(record)
        return {"statement_id": built.statement_id, "bytes": built.bytes}

    async def add_tcert(self, params: AddTcertParams) -> dict[str, Any]:
        if params.tcert_bytes:
            parsed = parse_signed_object(params.tcert_bytes)
            if parsed.type != "tcert":
                raise QrsValidationError("Provided object is not a TCert")
            tcert_id = tcert_id_of(parsed.signer_key_id, parsed.data["certificateNumber"])
            if tcert_id != params.target_tcert_id:
                raise QrsValidationError("Provided TCert bytes do not match the target tcert id")
            await self._deps.certificate_store.save(tcert_id, params.tcert_bytes)
        return await self.attest(AttestParams(
            ca_tcert_id=params.ca_tcert_id,
            target_tcert_id=params.target_tcert_id,
            claims=params.claims,
            issued_at=params.issued_at,
        ))

    async def resolve_trust(self, tcert_id: str, parsed: Any = None) -> TrustResolution:
        if await self._deps.trust_store.is_distrusted(tcert_id):
            return TrustResolution(state="invalid", pinned=False, message="TCert is locally distrusted")
        pinned = await self._deps.trust_store.is_pinned(tcert_id)
        attestations = await self._deps.trust_store.get_attestations(tcert_id)
        ca: dict[str, str | None] | None = None
        for att in attestations:
            if await self._is_valid_attestation(att):
                ca = {
                    "ca_tcert_id": att.ca_tcert_id,
                    "ca_name": att.claims.get("name") if att.claims else None,
                }
                break
        if pinned:
            return TrustResolution(state="valid", pinned=True, ca=ca)
        if ca:
            return TrustResolution(state="valid", pinned=False, ca=ca)
        return TrustResolution(
            state="cannotVerify",
            pinned=False,
            message="no trust path: TCert is neither pinned nor attested by a trusted CA",
        )

    async def is_trusted_ca(self, ca_tcert_id: str) -> bool:
        if await self._deps.trust_store.is_distrusted(ca_tcert_id):
            return False
        if await self._deps.revocation_store.get_revoked_tcert(ca_tcert_id):
            return False
        pinned = await self._deps.trust_store.is_pinned(ca_tcert_id)
        is_ca = await self._deps.trust_store.is_ca(ca_tcert_id)
        return pinned or is_ca

    async def _is_valid_attestation(self, att: AttestationRecord) -> bool:
        ca_bytes = await self._deps.certificate_store.get(att.ca_tcert_id)
        if not ca_bytes:
            return False
        try:
            ca_parsed = parse_signed_object(ca_bytes)
        except Exception:
            return False
        if ca_parsed.type != "tcert":
            return False
        provider = self._deps.crypto_registry.get(ca_parsed.algorithm)
        ca_pub = ca_parsed.data["publicKey"]
        if not verify_parsed_signed_object(ca_parsed, provider, ca_pub):
            return False
        if not await self.is_trusted_ca(att.ca_tcert_id):
            return False
        if await self._deps.revocation_store.get_revoked_key(att.ca_key_id):
            return False

        try:
            stmt = parse_statement(att.statement_bytes)
        except Exception:
            return False
        if stmt.action != "attest":
            return False
        if stmt.target.kind != "tcert":
            return False
        if tcert_id_of(stmt.target.key_id or "", stmt.target.certificate_number or 0) != att.target_tcert_id:
            return False
        if stmt.signer_key_id != att.ca_key_id:
            return False

        # The attestation must bind the exact content hash of the TCert it is about.
        target_bytes = await self._deps.certificate_store.get(att.target_tcert_id)
        if not target_bytes:
            return False
        try:
            target_parsed = parse_signed_object(target_bytes)
        except Exception:
            return False
        if tcert_hash_of(target_parsed) != att.tcert_hash:
            return False
        if stmt.target.tcert_hash != att.tcert_hash:
            return False

        return verify_statement(stmt.parsed, ca_pub, provider)

    async def _ensure_tcert(self, tcert_id: str) -> None:
        data = await self._deps.certificate_store.get(tcert_id)
        if not data:
            raise QrsNotFoundError(f"TCert not found: {tcert_id}")

    async def _key_info_of(self, tcert_id: str) -> dict[str, Any]:
        data = await self._deps.certificate_store.get(tcert_id)
        if not data:
            raise QrsNotFoundError(f"TCert not found: {tcert_id}")
        parsed = parse_signed_object(data)
        return {
            "key_id": parsed.signer_key_id,
            "algorithm": parsed.data["algorithm"],
            "public_jwk": parsed.data["publicKey"],
        }